### ISSUE TITLE
Gradle code generation for sql

### ISSUE THREAD

**Issue by Andro999b** [querydsl/querydsl#2770]
Is there is any clear documentation how to setup QueryDSL code generation for SQL/JDBC with gradle? I am only see option for maven and JPA.

**Comment by IncPlusPlus** [comment]
I had to poke around for a while so I get where you're coming from. I use querydsl for JPA entities in my project. You're using it for SQL/JDBC. Comparing the documentation, [the Maven config for JPA](http://www.querydsl.com/static/querydsl/latest/reference/html/ch02.html#d0e137) varies quite a bit from the [config for SQL/JDBC](http://www.querydsl.com/static/querydsl/latest/reference/html/ch02s03.html#d0e721). I don't have a project handy that I could try to configure to use JDBC but if it's any help, the content below is all I needed to add to my `build.gradle` to get Querysql working with my JPA entities.
```
dependencies {
    compile group: 'com.querydsl', name:'querydsl-jpa', version:'4.4.0'
    annotationProcessor group: 'com.querydsl', name: 'querydsl-apt', classifier: 'jpa', version: '4.4.0'
}
```
or you can do the condensed dependency notation:
```
dependencies {
    compile 'com.querydsl:querydsl-jpa:4.4.0'
    annotationProcessor 'com.querydsl:querydsl-apt:4.4.0'
}
```
[You might need to replace the use of "compile" in favor of "implementation"](https://docs.gradle.org/6.8/userguide/upgrading_version_5.html#dependencies_should_no_longer_be_declared_using_the_compile_and_runtime_configurations).

It seems like you'll also need the querydsl-maven-plugin. I don't know if there's a Gradle equivalent. There is a [Gradle plugin by a similar name](https://github.com/ewerk/querydsl-plugin) but it's not actually used for the same purpose as the maven plugin. [It was created for use before Gradle had native support for annotation processors](https://github.com/ewerk/gradle-plugins/issues/118#issuecomment-631303502). You might be out of luck in that regard.

**Comment by IncPlusPlus** [comment]
@Andro999b [According to feedback I received](https://github.com/querydsl/querydsl/issues/2772#issuecomment-780465334), the Maven plugin for querydsl-sql is used to generate the metamodel by querying the database for the schema (instead of having an annotation processor generate them from existing annotated classes). This means that the functionality the Maven plugin provides must be reproduced somehow.

The way I see it, to have these classes be generated you have three options. I've ranked these in difficulty from hardest to maintain to easiest in my opinion.
1. Write your own Gradle plugin that does what the Maven one does
2. Manually create your own 
3. Create a `pom.xml` that binds querydsl-maven-plugin to a specific lifecycle phase (`generate-sources` seems like an appropriate choice). Using the [gradle-maven-exec-plugin](https://github.com/dkorotych/gradle-maven-exec-plugin), you could then run a goal that covers the `generate-sources` phase.  There's also a chance you could execute the `querydsl-maven-plugin`'s `export` goal using [something like this](https://github.com/dkorotych/gradle-maven-exec-plugin/issues/28#issuecomment-765068142)

**Comment by cah-renan-ferreira** [comment]
Any solution on that?
I'm struggling to get queryDSL SQL in a gradle project =/

**Comment by IncPlusPlus** [comment]
@cah-renan-ferreira can you share the snippets of your `build.gradle` relevant to QueryDSL? This would include any dependencies or tasks involving it.

**Comment by renanleandrof** [comment]
Well, that's precisely the problem.
With maven, i use:
```
<plugin>
    <groupId>com.querydsl</groupId>
    <artifactId>querydsl-maven-plugin</artifactId>
    <version>5.0.0</version>
    <executions>
        <execution>
            <goals>
                <goal>export</goal>
            </goals>
        </execution>
    </executions>
    <configuration>
        <jdbcDriver>org.h2.Driver</jdbcDriver>
        <jdbcUrl>jdbc:h2:mem:CATALOGNAME;DATABASE_TO_UPPER=FALSE;INIT=RUNSCRIPT FROM 'file:src/main/resources/sql/schema.sql'</jdbcUrl>
        <packageName>io.github.crudzilla.example.sqlentities</packageName>
        <schemaToPackage>false</schemaToPackage>
        <namePrefix>S</namePrefix>
        <targetFolder>${project.basedir}/target/generated-sources/java</targetFolder>
        <sourceFolder/>
        <jdbcUser>sa</jdbcUser>
    </configuration>
    <dependencies>
        <dependency>
            <groupId>com.h2database</groupId>
            <artifactId>h2</artifactId>
            <version>1.4.200</version>
        </dependency>
    </dependencies>
</plugin>
```

But, with gradle, i have found no way to create the SQL entities =/

**Comment by IncPlusPlus** [comment]
Right. Gradle isn't officially supported. However, I've gotten it to work for me with just the dependency management features of Gradle. Luckily, I didn't need to create my own Gradle plugin. All it took for me was the following three lines.

```groovy
implementation group: 'com.querydsl', name:'querydsl-jpa', version:'+'
annotationProcessor group: 'com.querydsl', name: 'querydsl-apt', classifier: 'jpa', version: '+'
options.annotationProcessorGeneratedSourcesDirectory = file("${buildDir}/generated/src/main/java")
```

I'm using Spring Boot and Spring Data JPA. Because of that, my SQL entities are annotated with the `@javax.persistence.Entity` annotation from the [jakarta.persistence-api (JPA) package](https://search.maven.org/artifact/jakarta.persistence/jakarta.persistence-api/3.1.0/jar). Since you seem to be using H2 for your SQL, your configuration will be a little different. Here's a line-by-line breakdown.

The implementation line will need to have a different artifact name specified. I'd say you'd probably use [`querydsl-sql`](https://search.maven.org/artifact/com.querydsl/querydsl-sql/5.0.0/jar) instead.

The annotation processor line will require you to change the classifier. I had actually never seen the `classifier` property until recently. The way it's used is interesting. The `querydsl-apt` package is published with an additional item in the JAR's name. Instead of `artifact-name-1.2.3.jar`, they have artifact-name-1.2.3-classifier.jar`. Have a look at all the possible options [here](https://repo1.maven.org/maven2/com/querydsl/querydsl-apt/5.0.0/). My guess would be you'd either specify 'general' or omit the `classifier` property altogether.

Lastly, the `options.annotationProcessorGeneratedSourcesDirectory` was something I set so that Gradle would include the QueryDSL generated classes when compiling. You might not need this anymore so feel free to experiment. Hopefully this helps, @cah-renan-ferreira!

**Comment by renanleandrof** [comment]
I appreciate your help!
Tried these 3 lines and didn't worked for me.
First, i got a problem with this line:
options.annotationProcessorGeneratedSourcesDirectory = file("${buildDir}/generated/src/main/java")

Removing it, i got other error:

Execution failed for task ':compileJava'.
> java.lang.NoClassDefFoundError: javax/persistence/Entity

i'm using gradle 6.5

**Comment by IncPlusPlus** [comment]
Can you paste the exact 3 lines from your build.gradle?

**Comment by renanleandrof** [comment]
I copied exactly and added to first line in dependencies {...

**Comment by IncPlusPlus** [comment]
That's the problem. Like I said, _your dependencies will be different than mine_. Please read what I wrote below the three lines I pasted.

**Comment by renanleandrof** [comment]
Since I'm also gonna need the JPA entities classes, I'm trying first to
have exactly what you have, the JPA parts.

Then, I would proceed to try the sql entities.

Em sex., 13 de mai. de 2022 17:47, Ryan Cloherty ***@***.***>
escreveu:

**Comment by IncPlusPlus** [comment]
Now I'm confused. I never said you'd need the JPA dependency. I only pointed out that I was using JPA for my specific use-case. From what I understand, you're using H2 and SQL. If you're using the 'jpa' classifier for the annotation processor line, that's probably why you're having an error about being unable to find the `javax/persistence/Entity` class. You need to change the classifier like what I mentioned above.

**Comment by renanleandrof** [comment]
Would you mind to share your whole gradle build?

I'm using both JPA and SQL. At maven, I have 2 plug-ins, one to generate
the Q classes based on the JPA entities, and other (the one I shared) that
scans the h2 database to create the Sql entities, that I prefix with S.

Since you said you're successfuly using JPA, I would try to use your code
to first generate the JPA, and then try to get the other one.

Em sex., 13 de mai. de 2022 17:55, Ryan Cloherty ***@***.***>
escreveu:

**Comment by IncPlusPlus** [comment]
Ahhhhhh. I see where the confusion came from. I'm only using QueryDSL to generate Q classes from my JPA entities. Also, the project I use this for isn't open to the public so I can't share it.

I just realized why you're having the `NoClassDefFoundError`. It's because the JPA-related dependencies aren't in the runtime environment of the annotation processing stage. The workaround for this is to add whatever dependency you're using that introduces the JPA stuff as an annotation processor. Since the JPA classes are added to my project via a dependency on `spring-boot-starter-data-jpa`, I added the following to my build file.
```groovy
// see https://stackoverflow.com/a/52785316
// needed because the query dsl annotation processor doesn't recognize javax.persistence.Entity
// This puts the necessary dependencies on the classpath at annotation-processing time.
annotationProcessor group: 'org.springframework.boot', name: 'spring-boot-starter-data-jpa'
```
It's fine that `spring-boot-starter-data-jpa` isn't actually an annotation processor. It seems that Gradle will let you specify an ordinary library as an "annotation processor" and not cause a problem. It simply won't find any processors in the library and continue past it. However, it will also add the classes from that library into the runtime for all other annotation processors. That should fix the `NoClassDefFoundError` that occurs when `querydsl-apt` is scanning for the `@Entity` annotation.
